### ISSUE TITLE
Freeing the PureBasic IDE from all C Dependencies

### ISSUE THREAD

**Issue by tajmone** [fantaisie-software/purebasic#38]
**PROPOSAL** — Convert all C dependencies to PureBasic code and get rid of the need of C compilers in the build-chain. This is a _long term_ goal because it can't be accomplished until every C component is ported to PB.

**RATIONALE** — Having to set-up a C compiler and other C-related dependencies is a deterrent for many PB users who would otherwise like to work on the IDE project. This affects Windows users in particular, due to the need to install a full MSVS IDE (outdated version) and Win SDK, which are known to be problematic to install on modern Win versions (and this will only get worst as time passes).

Replacing the C modules with PureBasic code would also make cross-platform maintenance easier in many respects. Besides, having a native IDE written 100% in PureBasic code is a matter of pride in the language.

-------------------------------------------------------------------------------

## Looking into the C Dependencies

I've been sifting through the source code and trying to make heads and tails of what does what and where in the build chain (luckily, the comments in the sources are very useful in this respect). From what I gather, the C dependencies boil down to three distinct components:

- [ ]  [`PureBasicIDE/libmba/diff.c`][diff.c] (and its dependencies from __[libmba]__)
- [ ]  [`PureBasicIDE/AutomationDomainSocket.c`][Socket.c]
- [ ]  [`PureBasicDebugger/NetworkSupport.c`][Network.c]

The `diff.c` module from __[libmba]__ provides diffing functionality based on Eugene Myers' famous _O_(_ND_) diff algorithm (used by GNU diff), which the IDE uses to manage session history in [`PureBasicIDE/EditHistory.pb`][EditHistory.pb].

(for more info on __libmba__ and its role in this project, see [`PureBasicIDE/libmba/README.md`][libmba README]).

The other two modules deal with network functionality (especially sockets under Windows).

All of these modules should be re-implementable in PureBasic, although it might be a time-consuming task. My impression is that the diff algorithm is going to be easier part to deal with, whereas the network functions are going to be more challenging due to differences in native OS support, especially under Windows.

Also, in order to ensure that the PB-ported versions of these modules work as their C counter parts, it might be necessary to set-up some tests — e.g. a PB application that simulates their real-case usage and stress-tests both the original C modules and their PB versions, comparing the results for any discrepancies.

Although none of this is strictly necessary I think it's worth creating this Issue and putting forth the challenge.

On the one hand, one might argue that the trouble is not worth the result, and that having C components is just fine, especially if it makes life easier. My counter arguments to this are that **(1)** using C is certainly going to make life _uneasier_ for many PB users, and **(2)** I think that PureBasic's native IDE deserves the honour of being entirely written in PureBasic — if for anything, just for a matter of pride.

After all, we could also embrace the arguments in favour of adopting C components and take it to the extreme, arguing that it benefits from stronger compiler optimizations that are machine-specific, and so on, and then start to rewrite in C as many PB modules as possible (none of which would bring significant improvements in the light of what modern editors offer today). But my personal opinion is that such benefits would be marginal, not only in terms of performance or size (which on modern machines would be irrelevant) but especially if compared to the benefits of having an IDE which is 100% written in PureBasic, which would be accessible to _every_ PB user, without exception.

I think it's honest to say that most PB users love the language for the very reason that it lifts from them the burdens (and _horrors_) of having to deal with the C world, its Jurassic tools and infinite incompatible «standard» — and this applies also to those who do have to grapple with these horrors in their daily working life, for it's not a matter of competence that is at stake, but the love for _elegance_ and _simplicity_, and the desire to keep PureBasic assets in line with its core philosophy (which can be roughly resumed as "_let Fred handle the C dirty work that's behind the scenes!_").

At the cost of repeating myself, I think that the number of C components present in the IDE don't justify having to install the whole MS Visual Studio IDE and the Win SDK only to be able to compile the PureBasic IDE. We're speaking of just three components, none of which does anything transcendental — I think I've even seen that same diff algorithm implemented in PB before; and I've seen a full bit-torrent client written in PureBasic, so the network part should be doable too.

So we can look at this as a challenge to the PB community, as a (further) occasion to demonstrate that PB can stand on its feet and do just as well without C. And I'm not saying this has to be done today or tomorrow — it's just a long term proposal that would render the whole project more accessible to the PB user base, and that would also prevent issues with C dependencies becoming old in the course of time (some of them are already pretty legacy stuff under Windows).

[libmba]: http://www.ioplex.com/~miallen/libmba/ "Visit libmba home page"

[diff.c]: ../blob/master/PureBasicIDE/libmba/diff.c "View source file"
[Socket.c]: ../blob/master/PureBasicIDE/AutomationDomainSocket.c "View source file"
[Network.c]: ../blob/master/PureBasicDebugger/NetworkSupport.c "View source file"
[EditHistory.pb]: ../blob/master/PureBasicIDE/EditHistory.pb#L58 "View source file"
[libmba README]: ../blob/master/PureBasicIDE/libmba/README.md "View README file in libmba folder"

[xxx]: ../blob/master/

**Comment by NicolasGoeddel** [comment]
I think we shouldn't reinvent the wheel where it is not necessary and instead try to use all the libraries the operating system has to offer before somehow converting the functionality to Purebasic code. In the case of Linux you can simply install some additional dependencies and use the shared libraries to get the functionality you are looking for. For Windows you may find a DLL with the right functionality or maybe the OS already has it. On Mac I have no idea so far.
In my opinion the sense of shared libraries is to be able to get updated independently from the vendor of the main software. That is important in the case of possible security updates.

**Comment by tajmone** [comment]
@NicolasGoeddel

> I think we shouldn't reinvent the wheel where it is not necessary and instead try to use all the libraries the operating system has to offer before somehow converting the functionality to Purebasic code. 

This could be a viable solution for the network functionality, and I agree it would be better than implementing it in PureBasic, but it wouldn't cover the diff algorithm — that would require a PureBasic implementation, but as I said, that should be the easier one among all.

Do you have in mind any specific third party network library that could be used across all OSs to provide the required network functionality which is currently delegated to C? (If I remember correctly, the issue has to do with the need of using blocking socket functionality.)

Using a same library for all OSs might simplify maintenance.

> In the case of Linux you can simply install some additional dependencies and use the shared libraries to get the functionality you are looking for. For Windows you may find a DLL with the right functionality or maybe the OS already has it. On Mac I have no idea so far.

I wouldn't be so sure about Linux and macOS though, for Ubuntu and Apple are determined to drop support for 32-bit functionality — with macOS Catalina this is already happening, Ubuntu announced it months ago but then stepped back and postponed it after users complaints and Steam announcing that it will stop supporting games for Ubuntu if/when it dropped 32-bit support. 

So chances are that to build 32-bit PB apps on these systems will require some efforts, and even though it might still be possible to do so system DLLs will only be available for 64-bit soon, so relying on PureBasic code might be safer in case building 32-bit apps is still possible (for older macOS or other Linuxes). I'm not sure how that is going to unfold anyhow, probably it won't be possible to use 32-bit PB at all on those systems within a year time.

> In my opinion the sense of shared libraries is to be able to get updated independently from the vendor of the main software. That is important in the case of possible security updates.

Many PureLibraries are so old anyway that it would make little difference in the overall context ([see my post on PCRE status]). Besides, the current C modules for the network functionality were written specifically for the IDE, as for the diff from __libmba__ module it hasn't been updated since 2005 (in fact it had to be tweaked in a couple of places in order to compile, even with legacy versions of MSVS).

The root of the problem is that building the IDE requires legacy tools under Windows, which is a great obstacle for those who work on Win 10 and have installed the latest version of Visual Studio build tools and Win SDK. Therefore, unless pre-compiled binaries are included in the project, we need a solution that would allow end users to build the IDE focusing only on the PureBasic sources, without having to install huge packages just to build three small C components.



[see my post on PCRE status]: https://www.purebasic.fr/english/viewtopic.php?f=3&t=74037

**Comment by NicolasGoeddel** [comment]
> Do you have in mind any specific third party network library that could be used across all OSs to provide the required network functionality which is currently delegated to C? (If I remember correctly, the issue has to do with the need of using blocking socket functionality.)

At the moment I have no idea. What do you mean with "blocket socket functionality"? Normally sockets will block if the other side does not respond. Do we need [non-blocking sockets](https://www.scottklement.com/rpg/socktut/nonblocking.html) instead?

> I wouldn't be so sure about Linux and macOS though, for Ubuntu and Apple are determined to drop support for 32-bit functionality — with macOS Catalina this is already happening, Ubuntu announced it months ago but then stepped back and postponed it after users complaints and Steam announcing that it will stop supporting games for Ubuntu if/when it dropped 32-bit support.

I didn't think of 32-bit support. I can not remember when I had to use a 32-bit application the last time. I do not use Windows, but on Linux I have not a single application which runs on 32 bit. Is there really need for 32 bit? Which hardware do we want to support here? Even every single smartphone uses 64 bits nowadays.

> Many PureLibraries are so old anyway that it would make little difference in the overall context (see my post on PCRE status).

Ah, I remember. I also gave a +1 on that topic.

In general the Linux version of the IDE is the thing I am looking for. When I have a good amount of spare time I want to dig into its source code to get rid of the nasty autocompletion bug which prevents me to use the IDE. It's terrible it wasn't fixed since a year or even longer.

**Comment by tajmone** [comment]
> What do you mean with "blocket socket functionality"? Normally sockets will block if the other side does not respond. Do we need non-blocking sockets instead?

No, I meant exactly "_non-blocking_". I'm not sure about nitty gritty of all this, but I remember having read this in the C source of [`PureBasicDebugger/NetworkSupport.c`][NetworkSupport.c] (LL 18):

```c
/*
 ...
 *    We can't just use the PB network lib because it uses non-blocking sockets, also it won't
 *    work for receiving a thread becase the IDE and Debugger are not compiled threadsafe.
 ...
*/
```

... but there are other issues too. Fortunately all the sources are well commented in this regard (including failed attempts and after thoughts), so anyone willing to dig into it is going to find all the needed clues.

> Is there really need for 32 bit?

Of course there is, and the list of reasons is really long. Just to mention a few:

1. The x86_64 architecture is designed to run in 32-bit emulation mode, so there's really no good reason to lock-down the architecture potentiality via the OS.
2. I'm a great fan (and also an author) of Interactive Fiction text adventures. This is a legacy that goes back to '80s, with a huge number of tools and emulators that allow playing any adventure created from the '80s up to today. All these tools have only been built on 32-bit so far (for desktop PCs), and when macOS Catalina came out all the Mac users couldn't play or create any longer IF games!

    Now, when you think that this is the only video game genre that is truly accessible to the blind, it really sucks that overnight thousands of blind macOS users found themselves cut-off from these games — and for what reason?

    Right now there's a huge panic situation in the IF world regarding this, and since rebuilding all these tools for x64 might require ages (and probably their original developers are no longer maintaining them since decades) the only solution right now seem to be using commercial tools like [CrossOver], which was quick to seize the opportunity to [provide again 32-bit support] for Catalina users, or having to setup a VM to emulate a Win x64 OS, which still supports 32-bit! Right now, tools like Wine and other open source historical solution don't work anymore.
3. People wishing to learn Assembly should really start with 32-bit mode, which is easier for beginners to handle. This is going to hit hard on those PB users who invested energy into mixing PB and Assembly in their code.
4. All PureBasic 32-bit applications that used inline Assembly won't compiler any more. Again, why?
5. The sheer amount of legacy libraries (especially DLLs) that become unusable by locking out 32-bit from x64 OS is huge (and disgraceful). Just think of how many software that are available in both x32 and x64 end up being installed in x32 version just because there's a bigger echo system of plugins (especially audio tool).
5. 64-bit apps lead to memory bloat due to long pointers, which is not desirable in many types of applications. Not everyone thinks that the x64 architecture was such a great idea in hindsight. (if you're interested on the topic I can provide you some links to some interesting articles on the topic).

The list of reasons is really long...

> In general the Linux version of the IDE is the thing I am looking for.

The Linux build seems the one that gives less headaches.



[NetworkSupport.c]: https://github.com/fantaisie-software/purebasic/blob/master/PureBasicDebugger/NetworkSupport.c#L18
[CrossOver]: https://www.codeweavers.com/products/crossover-mac
[provide again 32-bit support]: https://liliputing.com/2019/12/crossover-19-runs-32-bit-windows-apps-on-64-bit-only-macos-catalina.html

**Comment by kenmo-pb** [comment]
I too would like to get rid of the C dependencies... *or* provide a compiler flag like `#USE_C_COMPONENTS = #False` which would allow people to work on the PB code and test it, while disabling the C requirements.

The C `diff` code could be ported to PB, or a simpler workaround is to always return `#False` from `History_MakeDiff()` when that compiler flag is set (I think this is the only usage of `libmba`?).

For the networking C code, I don't fully understand yet what it's used for - just the debugger? Could this be temporarily disabled via a compiler flag too? So that you can work on the IDE without including full debugger functionality? Need to investigate more.

**Comment by tajmone** [comment]
# Found PB Implementation of the Diff Algorithm

I've finally managed to undig from my notes the PureBasic port of the algorithm which I mentioned I had seen before...

This is the code, posted on the PB Forum by HeX0R:

- https://www.purebasic.fr/german/viewtopic.php?f=8&t=24147

The post is in German, so I can't understand it, but if I remember correctly this diff is part of HeX0R's __History Viewer__ IDE tool:

- https://www.purebasic.fr/english/viewtopic.php?f=27&t=45757
- http://hex0rs.coderbu.de/hv_linux
- http://hex0rs.coderbu.de/hv_win

Since the __History Viewer__ is closed source, we might need to contact the author to get permission to reuse the diff code from the forum (I vaguely remember contacting him asking to release the IDE tool sources but he denied the request).

The algorithm posted on the forum is the same one used by __libmba__ — i.e. Eugene Myers' famous O(ND) diff algorithm (also used by GNU diff). The PB code was ported from Matthias Hertel's C# implementation:

- http://www.mathertel.de/Diff/Default.aspx

So, if HeX0R authorizes us to reuse his port of the algorithm, half of the work is already done — we'd only need to adapt it to IDE code, in terms of interfacing and other issues.

**Comment by kenmo-pb** [comment]
I'm working on replacing the required `libmba` functions with all-PureBasic drop-in-compatible procedures. I think that's a good first step, just to drop the C dependency. Then later, `EditHistory.pb` can be further cleaned up, simplified, get rid of the `varray` functions and use PB native data structures, etc.

I have *not at all* been investigating the `NetworkSupport` functions...

**Comment by tajmone** [comment]
That's great news @kenmo-pb!

Just one thing, did anyone contact HeX0R about permission to reuse his code?

([clean room design] principle)

[clean room design]: https://en.wikipedia.org/wiki/Clean_room_design

**Comment by kenmo-pb** [comment]
For what it's worth, I haven't used HeX0R's code. I'm studying the basic Myers diff algorithm and implementing it specifically to the existing `libmba` / `EditHistory.pb` interface.

**Comment by tajmone** [comment]
> For what it's worth, I haven't used HeX0R's code. I'm studying the basic Myers diff algorithm and implementing it specifically to the existing libmba / EditHistory.pb interface.

That's really amazing, thanks @kenmo-pb! Please, if you don't mind, also publish the core algorithm elsewhere with a more permissive license than GPL (e.g. MIT) so that i might be used in other projects too. You could publish it as a GitHub Gist, so it doesn't require a repository but can still be version controlled and commented.

**Comment by t-harter** [comment]
If the diff dependency is solved, it should be rather simple to produce a C free version by disabling some features:

#### [`PureBasicIDE/AutomationDomainSocket.c`](../blob/master/PureBasicIDE/AutomationDomainSocket.c)
This file is only used on Linux where setting up the C build tools is much simpler. The Windows counterpart is message based and implemented directly in PB. 

It is also only needed for the incomplete and never used "Automation" feature (see [here](https://www.purebasic.fr/english/viewtopic.php?f=3&t=44895)) which was never fully completed. 

Maybe a good solution would be to simply remove that feature again.

#### [`PureBasicDebugger/NetworkSupport.c`](../blob/master/PureBasicDebugger/NetworkSupport.c)
This file is used for the "Network Debugging" feature of the standalone debugger. It is not used when compiled in the IDE. This functionality will not be easy to re-implement in PB since the binary layout of the socket functionality is quite different on each OS and hidden beneath a layer of C macros and definitions. The functionality must also match what the debugger is doing on the executable side and this part is naturally implemented in C. 

However, the Network Debugging is a rather isolated feature within the code. It would be simple to add compiler directives around it so it is possible to compile a version of the IDE/Debugger without this feature. This would simplify things especially for people who want to contribute on the IDE side and not the debugger anyways.

**Comment by kenmo-pb** [comment]
Good info @t-harter 

1. I didn't realize `AutomationDomainSocket.c` wasn't being compiled on Windows. Does that mean we're 1/3 done? 😁

2. For `NetworkSupport.c` I was thinking the same as you. It seems like an unneeded feature for just working on the IDE. So a flag to disable building it seems like a good idea (in the `Makefile` only? or also in the PB code?)

3. I have a working `diff.c` replacement in PB now (a "drop-in" replacement - it emulates the 4 functions imported from `libmba` - but not a direct code port). Let me test it more, clean it up, and I will submit it for review this week! It is *not optimized* for speed or memory, since it's my first time writing a basic Myers diff, so let's not merge directly into `master`.

**Comment by t-harter** [comment]
To exclude `NetworkSupport.c` there need to be some conditional compilation in the PB code too. Because much code is shared between IDE and debugger the IDE actually includes this file even if it is not really used. Its easy to do though by defining a PB constant from the makefile.

Regarding the diff replacement: 
Performance is not uncritical since the diff is used when storing history events which can happen for a lot of files on IDE shutdown when all files are closed. That is the reason why I went with an external library in the first place (and not with the simple/slow algorithm already implemented in `DiffWindow.pb`).

**Comment by tajmone** [comment]
> Performance is not uncritical since the diff is used when storing history events which can happen for a lot of files on IDE shutdown when all files are closed. That is the reason why I went with an external library in the first place

We could benchmark the performance differences between the PB implementation of the algorithm and the C version from libmba, and see if it's critical. Hopefully not. Also the PB version might have an advantage edge over the libmba version, for the former would be written specifically for the IDE needs, whereas the latter is more of a general purpose diff interface adapted to the IDE.

If the performance gap is not huge, I'd personally prefer to see the IDE wholly in PureBasic than relying on C libraries. Besides, there's always room for optimizations once the base algorithms is in place (and many optimizations to the base algorithm are available in the wild, and we could look into them later on). But I do understand your concerns for big projects involving many files (the IDE being a good example of a rather big project in itself).

**Comment by t-harter** [comment]
I agree. I just mentioned it since there were some complaints about the EditHistory performance in the past.

**Comment by kenmo-pb** [comment]
Good points @t-harter especially about closing/diffing all files at once. I have only tested some individual files.

I fully expect my PB version to be *slower* than `libmba`, because that uses an improved algorithm and surely deep-optimized code. Mine is a first attempt, but the performance difference could be negligible (milliseconds). Also I hope someone with more diff experience and/or C-to-PB experience can optimize it.

In general, maybe GitHub contributions can be included in PB Beta releases, for wider real-world testing? And easily undone if there are problems found.

**Comment by tajmone** [comment]
> In general, maybe GitHub contributions can be included in PB Beta releases, for wider real-world testing? And easily undone if there are problems found.

The idea that major/radical changes to the IDE could be first released as prototypes and tested by different users until everyone feels the IDE works fine sounds a good idea in general. I'm sure that announcing it on the forums should grant enough beta testers to allow developers to fine tune any problems via feedback, until the changes are considered safe to merge.

**Comment by kenmo-pb** [comment]
For what it's worth, my PB implementation of `diff` is pushed to a branch here:  
https://github.com/kenmo-pb/purebasic/tree/pb-diff-implementation

Set `#HISTORY_USE_NATIVE_DIFF` to `#True` in `EditHistory.pb`

Again, it's my first time ever implementing a Myers diff, so it's not very optimized for speed or memory, and the interface is a little strict because it's a drop-in replacement for the imported `libmba` functions.

But I'm questioning whether this really belongs in the official repo. As long as performance is critical (eg. releases), then `libmba` will probably be much better than this rough version. So this might be only useful for IDE developers, who don't want to set up the `libmba` dependencies... but in that case, we might as well just have a compile-time flag that disables Session History completely (or just disables the diff part).

**Comment by tajmone** [comment]
@kenmo-pb, thanks for the great work!

Did you have a chance to benchmark performance differences between your diff algorithm and the original **libmba**?

I've opened a new Issue on my fork were we can discuss how to create a C-free Win build, without cluttering the main repo:

- tajmone/purebasic/issues#4


I think it's worth doing some tests and see if we can come up with an alternative Windows solution to build the IDE without C dependencies — even if the PB diff algorithm is less performant, it would at least allow developers to actually build the IDE and test other features they might be working on.

Ideally, we could now create a PureBasic builder to compile the IDE for Windows, without requiring any external tool. I propose to carry on working on this feature in the other fork.

I'll do my best to join the work, but the situation right now in Italy is very tight due to the medical emergency which is affecting free time availability (simple things like daily shopping have become a struggle right now, slowing down everything).

**Comment by tajmone** [comment]
### Any Follow-Up With the PB Myer's Diff?

@kenmo-pb has done a great job with [his Myer's Diff PureBasic implementation], and his fork even provides [compiler directives to optionally enable building the IDE using the PB algorithm] instead of the C version from libmba.

Any chances that his PB diff algorithm could be incorporated in this repository, as an optional alternative to using the C version? 

This would enable more users to build the IDE without having to setup the whole C toolchain under Windows.

[his Myer's Diff PureBasic implementation]: https://github.com/kenmo-pb/purebasic/commit/b797784bbea79f476766419a2d72a3a8a67c52de "View Myer's Diff PureBasic implementation on @kenmo-pb fork"

[compiler directives to optionally enable building the IDE using the PB algorithm]: https://github.com/kenmo-pb/purebasic/blob/b797784bbea79f476766419a2d72a3a8a67c52de/PureBasicIDE/EditHistory.pb#L32 "View source variation on @kenmo-pb fork"

**Comment by kenmo-pb** [comment]
Thanks for the shoutout @tajmone .

I haven't worked on the native diff branch since sharing it. I think the best route would be to (1) implement further Myers improvements that are out there (2) do some comparison benchmarking vs `libmba` (3) seek some optimization gurus on the forums!

But I'll say again, if the goal is just to work on the IDE without `libmba`, maybe we just need a switch that disables the `diff` section or Session History. Probably a switch in the `makefile` too so it doesn't try to build `libmba`.

**Comment by tajmone** [comment]
@kenmo-pb I agree with all your proposals, especially offering alternative compiling options to enable building the IDE without C dependencies, even if it means dropping some features, since it would at least allow participation in the project for users (like me) who can't install the required dependencies.

Regarding alternative compilation, as @t-harter [mentioned earlier in this thread], it would be possible to add options to also disable compiling and using `AutomationDomainSocket.c` (for Linux) which is not used, and `NetworkSupport.c` which is not related to the IDE and could be skipped altogether in the make process.

Unfortunately, since I'm stuck with my dependencies problems, I'm not able to do any of the above changes and check that they don't break anything, because I can't compile the IDE on my machine (setting up a virtual machine is not an easy option, for various reasons, including the fact that I have an OEM Windows version which would not register in a VM). 

So I'm kind of stuck until the repository offers a C-less compiler toolchain — any changes I would do to the code would be untested, and potentially breaking things (as it already happened).

### Do We Really MSVS and Win7 SDK?

As a side note, I'm starting to wonder whether compiling the [libmba] Diff component _really_ requires setting up MSVS 2013 and the Win7 SDK, or whether it would simply compile using GCC under Windows. After all, the latest [libmba] update dates back to 2005 (Win7 was released in 2009) and the whole library was designed to be highly portable. 

So I'm starting to wonder whether the whole MSVS and Win7 SDK recommended setup is really needed for bulding just the IDE, or whether its requirements are for building the whole PureBasic package (i.e. for the language maintainers).

Anyhow, today I'll try to build the libmba component using GCC and see if it works. If it does, it would solve many issues for contributors, since GCC is a standalone compiler suite that only needs to be unzipped into a folder and added to the path (and a much better compiler than MSVS).

[mentioned earlier in this thread]: #issuecomment-576436063
[libmba]: http://www.ioplex.com/~miallen/libmba/

**Comment by kenmo-pb** [comment]
FYI :bulb: 

I am *not* saying this should be merged into the PureBasic repo, but it's here for **anybody who wants to get up and running with the IDE code quickly and easily**.
https://github.com/kenmo-pb/purebasic/tree/disable-c-dependencies  
It's a branch called `disable-c-dependencies` which lets you build and work on the IDE without the C setup/dependencies (at least on Windows).

- adds a `#DISABLE_C_DEPENDENCIES` switch in the PB code
- disables `libmba` (so Session History is disabled while you work on other parts of the IDE)
- disables `NetworkSupport` (rarely used Debugger feature?)
- removes those both from the IDE Makefile

Also, to avoid permission errors (writing to Program Files, etc.)

- disables `ColorTable` and `SilkTheme` in Makefile (assuming you aren't working on those)
- changes Makefile build dir from `HOME` to `TARGET`, a "local" subfolder

I believe you can clone and build this branch without MSVS. Feedback welcome. You still need `make` and UnxUtils, at least for now.

**Comment by tajmone** [comment]
Thanks @kenmo-pb, this is a great addition. It's unfortunate that this repository is still lacking a machine agnostic build script, but requires fiddling with the script in order to build the IDE — it's still more of a shared personal project that can be hacked to work on other machines.

**Comment by Naheulf** [comment]
```C
/*
 ...
 *    We can't just use the PB network lib because it uses non-blocking sockets, also it won't
 *    work for receiving a thread becase the IDE and Debugger are not compiled threadsafe.
 ...
*/
```

> #### [`PureBasicDebugger/NetworkSupport.c`](../blob/master/PureBasicDebugger/NetworkSupport.c)
> 
> This file is used for the "Network Debugging" feature of the standalone debugger. It is not used when compiled in the IDE. This functionality will not be easy to re-implement in PB since the binary layout of the socket functionality is quite different on each OS and hidden beneath a layer of C macros and definitions. The functionality must also match what the debugger is doing on the executable side and this part is naturally implemented in C.
> 
> However, the Network Debugging is a rather isolated feature within the code. It would be simple to add compiler directives around it so it is possible to compile a version of the IDE/Debugger without this feature. This would simplify things especially for people who want to contribute on the IDE side and not the debugger anyways.

I'm unable to understand the issue.

If the debugger require some blocking procedure unaviable in PB Network library,  can't a simple Delay loop like that do the job ?

```PureBasic
Procedure BlockingXXXXX(Args)
    Protected ReturnValue, Done = #False
    Repeat
        ReturnValue = OriginalXXXXX(Args)
        Done = CheckIfOriginalXXXXXJobIsDone() ; Can check if data-transfert is complete, connexion is established, ...
        If Not Done
            Delay(100)
        EndIf
    Until Done = #True
EndProcedure
```

On the other hand it seem that ReceiveNetworkData() wait for data before return (at least on Windows). So we can use something like that to have a non-blocking version if needed :
```PureBasic
Procedure NonBlockingReceiveNetworkData(ConnectionID, *Buffer, Size)
    Protected ReturnValue
    If NetworkClientEvent(ConnectionID) = #PB_NetworkEvent_Data
        ReturnValue = ReceiveNetworkData(ConnectionID, *Buffer, Size)
    EndIf
    ProcedureReturn ReturnValue
EndProcedure
```

Or maybe I missed some important thing...

**Comment by tajmone** [comment]
> If the debugger require some blocking procedure unaviable in PB Network library, can't a simple Delay loop like that do the job ?

If I remember correctly, the issue had more to do with thread safety.

In any case, the network support for the standalone debugger is being dropped (see PR #175), so I'm not sure if any of this is still going to be relevant in the nearby future.

**Comment by Naheulf** [comment]
> ## Looking into the C Dependencies
> 
> I've been sifting through the source code and trying to make heads and tails of what does what and where in the build chain (luckily, the comments in the sources are very useful in this respect). From what I gather, the C dependencies boil down to three distinct components:
> 
>     * [ ]   [`PureBasicIDE/libmba/diff.c`](../blob/master/PureBasicIDE/libmba/diff.c) (and its dependencies from **[libmba](http://www.ioplex.com/~miallen/libmba/)**)
> 
>     * [ ]   [`PureBasicIDE/AutomationDomainSocket.c`](../blob/master/PureBasicIDE/AutomationDomainSocket.c)
> 
>     * [ ]   [`PureBasicDebugger/NetworkSupport.c`](../blob/master/PureBasicDebugger/NetworkSupport.c)
> 

* `PureBasicIDE/libmba/diff.c` (and its dependencies from **libmba**) -> removed in #172 
* `PureBasicIDE/AutomationDomainSocket.c` -> removed in #169
* `PureBasicDebugger/NetworkSupport.c` -> removed in #175 

An from what I see in https://github.com/fantaisie-software/purebasic/network @t-harter is working on remove all re remaining C related things in https://github.com/fantaisie-software/purebasic/tree/simplify-build. And so that also will fix #35 in most parts.
